src/runtime_src/core/pcie/windows/shim.cpp

```diff
@@ -1610,3 +1610,387 @@ xclReadTraceData(xclDeviceHandle handle, void* traceBuf, uint32_t traceBufSz,
+  };
+
+  uint64_t baseAddress[XLAPC_MAX_NUMBER_SLOTS];
+  uint32_t numSlots = getDebugIpData(handle, LAPC, baseAddress, nullptr, nullptr, nullptr, nullptr, XLAPC_MAX_NUMBER_SLOTS);
+  uint32_t temp[XLAPC_STATUS_PER_SLOT];
+  aCheckerResults->NumSlots = numSlots;
+  //snprintf(aCheckerResults->DevUserName, 256, "%s", mDevUserName.c_str());
+  for (uint32_t s = 0; s < numSlots; ++s) {
+    for (int c=0; c < XLAPC_STATUS_PER_SLOT; c++)
+      size += shim->read(XCL_ADDR_SPACE_DEVICE_CHECKER, baseAddress[s]+statusRegisters[c], &temp[c], 4);
+
+    aCheckerResults->OverallStatus[s] = temp[XLAPC_OVERALL_STATUS];
+    std::copy(temp+XLAPC_CUMULATIVE_STATUS_0, temp+XLAPC_SNAPSHOT_STATUS_0, aCheckerResults->CumulativeStatus[s]);
+    std::copy(temp+XLAPC_SNAPSHOT_STATUS_0, temp+XLAPC_STATUS_PER_SLOT, aCheckerResults->SnapshotStatus[s]);
+  }
+  return size;
+}
+
+
+// Read APM performance counters
+size_t xclDebugReadCounters(xclDeviceHandle handle, xclDebugCountersResults* aCounterResults) 
+{
+  auto shim = get_shim_object(handle);
+  size_t size = 0;
+
+  uint64_t spm_offsets[] = {
+    XAIM_SAMPLE_WRITE_BYTES_OFFSET,
+    XAIM_SAMPLE_WRITE_TRANX_OFFSET,
+    XAIM_SAMPLE_READ_BYTES_OFFSET,
+    XAIM_SAMPLE_READ_TRANX_OFFSET,
+    XAIM_SAMPLE_OUTSTANDING_COUNTS_OFFSET,
+    XAIM_SAMPLE_LAST_WRITE_ADDRESS_OFFSET,
+    XAIM_SAMPLE_LAST_WRITE_DATA_OFFSET,
+    XAIM_SAMPLE_LAST_READ_ADDRESS_OFFSET,
+    XAIM_SAMPLE_LAST_READ_DATA_OFFSET
+  };
+
+  uint64_t spm_upper_offsets[] = {
+    XAIM_SAMPLE_WRITE_BYTES_UPPER_OFFSET,
+    XAIM_SAMPLE_WRITE_TRANX_UPPER_OFFSET,
+    XAIM_SAMPLE_READ_BYTES_UPPER_OFFSET,
+    XAIM_SAMPLE_READ_TRANX_UPPER_OFFSET,
+    XAIM_SAMPLE_OUTSTANDING_COUNTS_UPPER_OFFSET,
+    XAIM_SAMPLE_LAST_WRITE_ADDRESS_UPPER_OFFSET,
+    XAIM_SAMPLE_LAST_WRITE_DATA_UPPER_OFFSET,
+    XAIM_SAMPLE_LAST_READ_ADDRESS_UPPER_OFFSET,
+    XAIM_SAMPLE_LAST_READ_DATA_UPPER_OFFSET
+  };
+
+  // Read all metric counters
+  uint64_t baseAddress[XAIM_MAX_NUMBER_SLOTS];
+  uint8_t  perfMonProperties[XAIM_MAX_NUMBER_SLOTS] = {};
+  uint32_t numSlots = getDebugIpData(handle, AXI_MM_MONITOR, baseAddress, nullptr, perfMonProperties, nullptr, nullptr, XAIM_MAX_NUMBER_SLOTS);
+
+  uint32_t temp[XAIM_DEBUG_SAMPLE_COUNTERS_PER_SLOT];
+
+  aCounterResults->NumSlots = numSlots;
+  
+  for (uint32_t s=0; s < numSlots; s++) {
+    uint32_t sampleInterval;
+    // Read sample interval register to latch the sampled metric counters
+    size += shim->read(XCL_ADDR_SPACE_DEVICE_PERFMON,
+                    baseAddress[s] + XAIM_SAMPLE_OFFSET,
+                    &sampleInterval, 4);
+
+    // If applicable, read the upper 32-bits of the 64-bit debug counters
+    if (perfMonProperties[s] & XAIM_64BIT_PROPERTY_MASK) {
+      for (int c = 0 ; c < XAIM_DEBUG_SAMPLE_COUNTERS_PER_SLOT ; ++c) {
+        shim->read(XCL_ADDR_SPACE_DEVICE_PERFMON,
+                   baseAddress[s] + spm_upper_offsets[c], &temp[c], 4) ;
```

Comment:
Never good to have "magic" numbers.  In this case, replace the 4 with sizeof(uint32_t)